src/prelude/internals.mo

```diff
@@ -435,3 +435,151 @@ func @create_actor_helper(wasm_module_ : Blob, arg_ : Blob) : async Principal =
 };
+
+// default timer mechanism implementation
+// fundamental node invariant: max_exp pre <= expire <= min_exp post
+// corollary: if expire == 0 then the pre is completely expired
+//
+// Note: Below the `expire` field is an encoding of an aliased mutable field with
+//       a single-element mutable array. It eliminates `--experimental-field-aliasing`
+//       while compiling this file at the cost of slightly higher syntactic noise
+//       as well as increased allocation and runtime cost accessing the data. Oh well.
+//
+type @Node = { expire : [var Nat64]; id : Nat; delay : ?Nat64; job : () -> async (); pre : ?@Node; post : ?@Node };
+var @timers : ?@Node = null;
+func @prune(n : ?@Node) : ?@Node = switch n {
+  case null null;
+  case (?n) {
+    if (n.expire[0] == 0) {
+      @prune(n.post) // by corollary
+    } else {
+      ?{ n with pre = @prune(n.pre); post = @prune(n.post) }
+    }
+  }
+};
+
+func @nextExpiration(n : ?@Node) : Nat64 = switch n {
+  case null 0;
+  case (?n) {
+    var exp = @nextExpiration(n.pre); // TODO: use the corollary for expire == 0
+    if (exp == 0) {
+      exp := n.expire[0];
+      if (exp == 0) {
+        exp := @nextExpiration(n.post)
+      }
+    };
+    exp
+  }
+};
+
+// Function called by backend to run eligible timed actions.
+// DO NOT RENAME without modifying compilation.
+func @timer_helper() : async () {
+  func Array_init<T>(len : Nat,  x : T) : [var T] {
+    (prim "Array.init" : <T>(Nat, T) -> [var T])<T>(len, x)
+  };
+  let now = (prim "time" : () -> Nat64)();
+  let exp = @nextExpiration @timers;
+  let prev = (prim "global_timer_set" : Nat64 -> Nat64) exp;
+
+  if (exp == 0) {
+    return
+  };
+
+  var gathered = 0;
+  let thunks = Array_init<?(() -> async ())>(10, null); // we want max 10
+
+  func gatherExpired(n : ?@Node) = switch n {
+    case null ();
+    case (?n) {
+      gatherExpired(n.pre);
+      if (n.expire[0] > 0 and n.expire[0] <= now and gathered < thunks.size()) {
+        thunks[gathered] := ?(n.job);
+        switch (n.delay) {
+          case (?delay) if (delay != 0) {
+            // re-add the node
+            let expire = n.expire[0] + delay;
+            // N.B. insert only works on pruned nodes
+            func reinsert(m : ?@Node) : @Node = switch m {
+              case null ({ n with expire = [var expire]; pre = null; post = null });
+              case (?m) {
+                assert m.expire[0] != 0;
+                if (expire < m.expire[0]) ({ m with pre = ?reinsert(m.pre) })
+                else ({ m with post = ?reinsert(m.post) })
+              }
+            };
+            @timers := ?reinsert(@prune(@timers));
+          };
+          case _ ()
+        };
+        n.expire[0] := 0;
+        gathered += 1;
+      };
+      gatherExpired(n.post);
+    }
+  };
+
+  gatherExpired(@timers);
+
+  for (k in thunks.keys()) {
+    ignore switch (thunks[k]) { case (?thunk) ?thunk(); case _ null };
```

Comment:
@crusso So what happens if this `throw`s? maybe we can just run the `jobs on the fly while gathering and `catch` when the queue is full. Otherwise expunge.

Reply:
Yes, I think you could do that - no need for the array and second traversal then.

Reply:
I agree but we are not there yet. I leave tis to a follow-up PR.